pkg/shared/service.go

```diff
@@ -17,7 +19,139 @@ func NewServiceHandler() *ServiceHandler {
 }
 
 func (sh *ServiceHandler) Handle(ctx context.Context, event sdk.Event) error {
-	fmt.Println("handling object ", event.Object.GetObjectKind().GroupVersionKind().String())
+	logrus.Debug("handling object ", event.Object.GetObjectKind().GroupVersionKind().String())
+
+	sharedService := event.Object.(*v1alpha1.SharedService)
+	sharedServiceCopy := sharedService.DeepCopy()
+
+	if sharedService.Spec.ServiceType != strings.ToLower(v1alpha1.KeycloakKind) {
+		return nil
+	}
+
+	if event.Deleted {
+		return nil
+	}
+
+	if sharedService.GetDeletionTimestamp() != nil {
+		return sh.finalizeSharedService(sharedServiceCopy)
+	}
+
+	logrus.Debugf("SharedServicePhase: %v", sharedService.Status.Phase)
+	switch sharedService.Status.Phase {
+	case v1alpha1.SSPhaseNone:
+		sh.initSharedService(sharedServiceCopy)
+	case v1alpha1.SSPhaseAccepted:
+		sh.createKeycloaks(sharedServiceCopy)
+	}
+
+	return nil
+}
+
+func (sh *ServiceHandler) createKeycloaks(sharedService *v1alpha1.SharedService) error {
+	keycloakList := v1alpha1.KeycloakList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Keycloak",
+			APIVersion: "aerogear.org/v1alpha1",
+		},
+	}
+	listOptions := sdk.WithListOptions(&metav1.ListOptions{
+		LabelSelector:        "aerogear.org/sharedServiceName=" + sharedService.ObjectMeta.Name,
+		IncludeUninitialized: false,
+	})
+
+	err := sdk.List(sharedService.Namespace, &keycloakList, listOptions)
+	if err != nil {
+		logrus.Errorf("Failed to query keycloaks : %v", err)
+		return err
+	}
+
+	numCurrentInstances := len(keycloakList.Items)
+	numRequiredInstances := minRequiredInstances(sharedService.Spec.MinInstances, sharedService.Spec.RequiredInstances)
+	numMaxInstances := sharedService.Spec.MaxInstances
+	logrus.Debugf("number of service instances(%v), current: %v, required: %v, max: %v", sharedService.Spec.ServiceType, numCurrentInstances, numRequiredInstances, numMaxInstances)
+
+	if numCurrentInstances < numRequiredInstances {
+		err := sdk.Create(newKeycloak(sharedService))
+		if err != nil {
+			logrus.Errorf("Failed to create keycloak : %v", err)
+			return err
+		}
+	} else {
+		sharedService.Status.Ready = true
+		sharedService.Status.Phase = v1alpha1.SSPhaseComplete
+		err := sdk.Update(sharedService)
+		if err != nil {
+			logrus.Errorf("error updating resource status: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func minRequiredInstances(minInstances, requiredInstances int) int {
+	if minInstances > requiredInstances {
+		return minInstances
+	}
+	return requiredInstances
+}
+
+func newKeycloak(sharedService *v1alpha1.SharedService) *v1alpha1.Keycloak {
+	labels := map[string]string{
+		"aerogear.org/sharedServiceName": sharedService.ObjectMeta.Name,
+	}
```

Comment:
Added this label to all keycloaks so that we can use the operator sdks List methods using label selectors to get the list of keycloaks for this share service.